BFS_Connected_Components/Q10_Leetcode_number_of_closed_islands.py

```diff
@@ -0,0 +1,38 @@
+import collections
+def closedIsland(grid):
+    ROWS, COLS = len(grid), len(grid[0])
+    
+    directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
+    seen = set()
+    
+
+    def is_closed_island(r, c):
+        q = collections.deque([(r, c)])
+        seen.add((r, c))
+        closed = True
+        while q:
+            row, col = q.popleft()
+            is_boundary = row == ROWS - 1 or row == 0 or col == 0 or col == COLS - 1
+            if is_boundary:
+                    closed = False
+            for dr, dc in directions:
+                r, c = row + dr, col + dc
+                is_in_bounds = r in range(ROWS) and c in range(COLS)
```

Comment:
The statement `r in range(ROWS)` is O(ROWS) time. Instead, you should use a check like `row >= 0 and row < num_rows`. For the sake of reusability, it would be nice to create a separate function to check if a given position is in bounds.